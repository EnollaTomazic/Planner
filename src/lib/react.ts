import * as React from "react";

export const hasTextContent = (node: React.ReactNode): boolean => {
  if (node === null || node === undefined) return false;
  if (typeof node === "boolean") return false;
  if (typeof node === "string") return node.trim().length > 0;
  if (typeof node === "number") return true;
  if (Array.isArray(node)) return node.some((item) => hasTextContent(item));
  if (React.isValidElement<{ children?: React.ReactNode }>(node)) {
    return hasTextContent(node.props.children ?? null);
  }
  return false;
};

export function observeMediaQuery(
  query: string,
  onChange: (matches: boolean, mediaQuery: MediaQueryList) => void,
): (() => void) | undefined {
  if (typeof window === "undefined" || typeof window.matchMedia !== "function") {
    return undefined;
  }

  let mediaQuery: MediaQueryList;

  try {
    mediaQuery = window.matchMedia(query);
  } catch {
    return undefined;
  }

  const handleChange = (event: MediaQueryListEvent): void => {
    onChange(event.matches, mediaQuery);
  };

  const handleLegacyChange = (): void => {
    onChange(mediaQuery.matches, mediaQuery);
  };

  onChange(mediaQuery.matches, mediaQuery);

  if (typeof mediaQuery.addEventListener === "function") {
    mediaQuery.addEventListener("change", handleChange);
    return () => {
      mediaQuery.removeEventListener("change", handleChange);
    };
  }

  if (typeof mediaQuery.addListener === "function") {
    mediaQuery.addListener(handleLegacyChange);
    return () => {
      mediaQuery.removeListener(handleLegacyChange);
    };
  }

  return undefined;
}

export function useMatchMedia(
  query: string,
  getServerSnapshot?: () => boolean,
): boolean {
  const getMatches = React.useCallback(() => {
    if (typeof window === "undefined") {
      return typeof getServerSnapshot === "function" ? getServerSnapshot() : false;
    }

    if (typeof window.matchMedia !== "function") {
      return false;
    }

    try {
      return window.matchMedia(query).matches;
    } catch {
      return false;
    }
  }, [query, getServerSnapshot]);

  const [matches, setMatches] = React.useState<boolean>(getMatches);

  React.useEffect(() => {
    return observeMediaQuery(query, (nextMatches) => {
      setMatches(nextMatches);
    });
  }, [query]);

  return matches;
}
