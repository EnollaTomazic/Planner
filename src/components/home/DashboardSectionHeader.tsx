"use client";

import * as React from "react";

import { cn } from "@/lib/utils";
import { Skeleton } from "@/components/ui";

export interface DashboardSectionHeaderShowCodeProps {
  id?: string;
  controls?: string;
  expanded?: boolean;
  disabled?: boolean;
  onToggle?: () => void;
  label?: string;
  expandedLabel?: string;
}

export interface DashboardSectionHeaderProps {
  title: string;
  actions?: React.ReactNode;
  loading?: boolean;
  className?: string;
  showCode?: DashboardSectionHeaderShowCodeProps;
}

const baseActionGap =
  "flex items-center gap-[var(--space-2)] sm:gap-[var(--space-3)]";

function renderTitle(title: string, loading?: boolean) {
  if (loading) {
    return (
      <Skeleton
        ariaHidden={false}
        role="status"
        aria-label="Loading section title"
        radius="sm"
        className="!h-[var(--space-5)] !w-[min(60%,var(--space-16))]"
      />
    );
  }

  return (
    <h2 className="text-title font-semibold tracking-[-0.01em] text-card-foreground">
      {title}
    </h2>
  );
}

function renderShowCode(
  showCode: DashboardSectionHeaderShowCodeProps | undefined,
  loading?: boolean,
) {
  if (!showCode && !loading) {
    return null;
  }

  if (loading) {
    return (
      <Skeleton
        ariaHidden={false}
        role="status"
        aria-label="Loading Show code action"
        radius="full"
        className="!h-[var(--control-h-sm)] !w-[calc(var(--space-8)*2)]"
      />
    );
  }

  if (!showCode) {
    return null;
  }

  const {
    id,
    controls,
    expanded,
    disabled,
    onToggle,
    label = "Show code",
    expandedLabel = "Hide code",
  } = showCode;

  return (
    <button
      type="button"
      id={id}
      onClick={onToggle}
      aria-controls={controls}
      aria-expanded={expanded}
      disabled={disabled}
      data-state={expanded ? "expanded" : undefined}
      className={cn(
        "inline-flex items-center gap-[var(--space-1)] rounded-[var(--radius-md)] px-[var(--space-2)] py-[var(--space-1)]",
        "text-label font-medium text-accent-3 underline underline-offset-[var(--space-1)]",
        "transition-colors focus-visible:outline-none focus-visible:ring-[var(--ring-size-1)]",
        "focus-visible:ring-offset-0 ring-[var(--theme-ring)]",
        "disabled:cursor-not-allowed disabled:opacity-disabled",
      )}
    >
      {expanded ? expandedLabel : label}
    </button>
  );
}

export function DashboardSectionHeader({
  title,
  actions,
  loading,
  className,
  showCode,
}: DashboardSectionHeaderProps) {
  const shouldRenderActions = Boolean(actions) || Boolean(showCode) || loading;

  return (
    <div
      className={cn(
        "flex flex-wrap items-center justify-between gap-y-[var(--space-2)] gap-x-[var(--space-3)]",
        className,
      )}
    >
      {renderTitle(title, loading)}
      {shouldRenderActions ? (
        <div
          className={cn(
            baseActionGap,
            "sm:justify-end",
            loading ? "w-full sm:w-auto" : undefined,
          )}
        >
          {actions}
          {renderShowCode(showCode, loading)}
        </div>
      ) : null}
    </div>
  );
}
