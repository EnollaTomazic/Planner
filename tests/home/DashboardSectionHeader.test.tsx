import * as React from "react";
import { render } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";

import DashboardSectionHeader from "@/components/home/DashboardSectionHeader";

describe("DashboardSectionHeader", () => {
  it("renders the title layout with actions and show code button", () => {
    const toggleSpy = vi.fn();
    const { container } = render(
      <DashboardSectionHeader
        title="Planner"
        actions={<button type="button">Refresh</button>}
        showCode={{
          controls: "planner-code",
          onToggle: toggleSpy,
          expanded: false,
        }}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders skeleton placeholders while loading", () => {
    const { container } = render(
      <DashboardSectionHeader title="Loading state" loading />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
