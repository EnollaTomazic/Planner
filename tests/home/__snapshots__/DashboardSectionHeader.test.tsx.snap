// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DashboardSectionHeader > renders skeleton placeholders while loading 1`] = `
<div
  class="flex flex-wrap items-center justify-between gap-y-[var(--space-2)] gap-x-[var(--space-3)]"
>
  <div
    aria-hidden="false"
    aria-label="Loading section title"
    class="skeleton h-[var(--space-4)] w-full rounded-[var(--radius-sm,var(--radius-md))] !h-[var(--space-5)] !w-[min(60%,var(--space-16))]"
    role="status"
  />
  <div
    class="flex items-center gap-[var(--space-2)] sm:gap-[var(--space-3)] sm:justify-end w-full sm:w-auto"
  >
    <div
      aria-hidden="false"
      aria-label="Loading Show code action"
      class="skeleton h-[var(--space-4)] w-full rounded-[var(--radius-full)] !h-[var(--control-h-sm)] !w-[calc(var(--space-8)*2)]"
      role="status"
    />
  </div>
</div>
`;

exports[`DashboardSectionHeader > renders the title layout with actions and show code button 1`] = `
<div
  class="flex flex-wrap items-center justify-between gap-y-[var(--space-2)] gap-x-[var(--space-3)]"
>
  <h2
    class="text-title font-semibold tracking-[-0.01em] text-card-foreground"
  >
    Planner
  </h2>
  <div
    class="flex items-center gap-[var(--space-2)] sm:gap-[var(--space-3)] sm:justify-end"
  >
    <button
      type="button"
    >
      Refresh
    </button>
    <button
      aria-controls="planner-code"
      aria-expanded="false"
      class="inline-flex items-center gap-[var(--space-1)] rounded-[var(--radius-sm)] px-[var(--space-2)] py-[var(--space-1)] text-label font-medium text-accent-3 underline underline-offset-[var(--space-1)] transition-colors focus-visible:outline-none focus-visible:ring-[var(--ring-size-1)] focus-visible:ring-offset-0 ring-[var(--theme-ring)] disabled:cursor-not-allowed disabled:opacity-disabled"
      type="button"
    >
      Show code
    </button>
  </div>
</div>
`;
